sweep/sweeper.go

```diff
@@ -1521,12 +1393,76 @@ func (s *UtxoSweeper) handleNewInput(input *sweepInputMessage,
 	)
 	if err != nil {
 		err := fmt.Errorf("wait for spend: %w", err)
-		s.signalAndRemove(&outpoint, Result{Err: err})
+		s.markInputFailed(pi, err)
 
 		return
 	}
 
-	pendInput.ntfnRegCancel = cancel
+	pi.ntfnRegCancel = cancel
+}
+
+// decideStateAndRBFInfo queries the mempool to see whether the given input has
+// already been spent. If so, the state Published will be returned, otherwise
+// state Init. When spent, it will query the sweeper store to fetch the fee
+// info of the spending transction, and construct an RBFInfo based on it.
+// Suppose an error occurs, fn.None is returned.
+func (s *UtxoSweeper) decideStateAndRBFInfo(op wire.OutPoint) (
```

Comment:
So the basic idea for this function is that it allows us to recover RBF info after a restart, right?

I wonder if we should also use this function to detect a competing spend from another party.  For example, suppose we have an HTLC-Timeout / preimage spend bidding war going on.  It would be good to detect the other party's fee/rate so we can replace their transaction in the mempool.

Or does the "fee bumper" appropriately handle this case?

Reply:
> So the basic idea for this function is that it allows us to recover RBF info after a restart, right?

Exactly. And the rest should be handled by the fee bumper - however, the whole recovering-from-restart is not implemented in these PR series in order to limit the scope. With the existing design, I think it should be straightforward - we just need to grab the inputs here, make them into a set, and re-offer it to the fee bumper.

As for the race case, it's more complicated. My general approach is to avoid it as much as possible, by extracting the preimage from the mempool, #7564 helps, tho it'd be nicer if we could also have #7615. As for the role of fee bumper, it just blindly does RBF when a new block comes in, and will use up all the budget configured by the user by the time the race happens.

Reply:
> For example, suppose we have an HTLC-Timeout / preimage spend bidding war going on. It would be good to detect the other party's fee/rate so we can replace their transaction in the mempool.

So for that, we created this PR (https://github.com/lightningnetwork/lnd/pull/7564) which allows resolvers to directly get notified if a spend of a transaction happens in the mempool.